### ISSUE TITLE
Cached data from current release occasionally causing new releases to fail

### ISSUE THREAD

**Issue by sjb9774** [davidalger/capistrano-magento2#138]
# Summary
In 2.3.4 and 2.3.5 there have been a number of deployments I've run where the deployment fails (usually during setup:upgrade) with Magento complaining that some class/table/configuration doesn't. After thoroughly debugging one of these scenarios I've come up with the following explanation that I believe makes sense. 

1. During a deployment, `bin/magento maintenance:enable` is often run. It is executed from the directory of the existing deployment
1. I believe during the execution of this command, the Magento cache (Redis in our case) is populated with information loaded from module config files (`indexer.xml` for example)
1. When later commands are run from the new release directory, if the configuration that was cached during the execution of the `maintenance:enable` command conflicts with the information expected in the new release, you may encounter errors
1. Enabling maintenance mode _before_ beginning the deployment can circumvent this problem, as the cache doesn't get populated from within `current`

## Example
For a concrete example, I'll describe the most recent scenario I encountered:
1. Magento Commerce 2.3.4 client wishes to disable the MSI modules
1. A commit is created, disabling all the relevant modules in `config.php`
1. Upon deployment, an error is generated during `setup:db-data:upgrade` 
``` magento stderr: In Mysql.php line 110:
  SQLSTATE[42S02]: Base table or view not found: 1146 Table 'mymagento.inventory_source_item' doesn't exist, query was: CREATE TRIGGER trg_inventory_source_item_after_insert AFTER INSERT ON inventory_source_item FOR EACH ROW BEGIN INSERT IGNORE INTO `inventory_cl` (`entity_id`) VALUES (NEW.`source_item_id`); END
```
1. Investigating the related code revealed that ultimately this was because the `\Magento\Framework\Indexer\ConfigInterface::getIndexers` within `\Magento\Indexer\Setup\RecurringData::install` was returning the `inventory` indexer from `vendor/magento/module-inventory-indexer/etc/indexer.xml` despite that module being disabled
1. This issue was not reproducible locally
1. On the server, in the current deployment directory I used the `mr dev:console` to run `\Magento\Framework\Indexer\ConfigInterface::getIndexers` and verified that the `inventory` indexer returned. This makes sense, as the previous deployment had failed so the InventoryIndexer module was still enabled
1. I moved to the release directory of the failed deployment and again used `mr dev:console` and ran `\Magento\Framework\Indexer\ConfigInterface::getIndexers` confirmed that the `inventory` indexer _was still present despite being disabled`
1. I then ran a cache flush and retried the `dev:console` command, and was able to confirm that the `inventory` indexer no longer appeared
1. A subsequent redeploy continued without error

## Solution?
I'm not sure what a proper fix for this looks like (perhaps a separate cache to be used for release directories?), or if it perhaps even is more the responsibility of Magento to not populate the cache during `maintenance:enable` but I did want to document this issue as I've encountered it three times on different sites in as many weeks.

**Comment by davidalger** [comment]
@sjb9774 Thanks for the stellar write up! A colleague of mine, @vbuck has also encountered this issue, and through experimentation resolved it on the affected project by adding the following to `lib/capistrano/tasks/pre_upgrade.rake`
```
before 'magento:app:config:import', 'magento:setup:flush_cache' do
  on cache_hosts do
    within release_path do
      execute :magento, 'cache:flush'
    end
  end
end
```

I believe under the hood Magento is now expecting the cache to be flushed before app:config:import is run due to certain scenarios like this.

An alternative workaround is to dump i18n, themes and scopes information to the `app/etc/config.php` file which allows zero-down deployments to be executed in the majority of cases. But what that also does, is pull the env.php out of the current release while it's being built preventing any poisoning of the cache, and then re-links it later on before putting the release live.

Fixing this I believe merely amounts to adding the cache:flush call immediately prior to app:config:import when maintenance is required for the upgrade routine. You've caught me at a good time for this, as I'm working on cutting a 0.9.0 release, and will include this fix in said release. Again, thanks for the report. Appreciate it.

**Comment by davidalger** [comment]
Published in version 0.9.0 of this gem. See changelog for full details of that release.

**Comment by sjb9774** [comment]
Great to hear @davidalger, thanks for the prompt response!
